### PR TITLE
chore(docs): add python api protection code

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

add python api protection code

image

image

## Testing

test locally
@logto-io/eng

### CODE REVIEW

File: docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx [hunk 1]
```diff
@@ -56,5 +56,39 @@ public class JwtRequestFilter extends OncePerRequestFilter {
...
+# For django developers:
+from authlib.integrations.django_oauth2 import ResourceProtector
+
+# For Flask Developers
+from authlib.integrations.flask_oauth2 import ResourceProtector
+
+# We will get back to this validator customization on the following validation section
+from validator import MyLogtoBearerTokenValidator
+
+APP = Flask(__name__)
```
Comment (gao-sun) [inline docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx:82]: Is `APP` by design? or just
```suggestion
App = Flask(__name__)
```

File: docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx [hunk 2]
```diff
@@ -128,5 +128,56 @@ Package selections vary depending on the framework you are using.
...
+    """
+    jwks: You will need to inquire the jwks from Logto's jwks endpoint: "https://<your-logto-domain>/oidc/jwks" and pass in the response
+    """
+    def __init__(self, jwks, issuer="https://<your-logto-domain>/oidc", audience):
+        public_key = JsonWebKey.import_key_set(jwks)
+        super(MyLogtoBearerTokenValidator, self).__init__(public_key, issuer)
+        self.claims_options = {
+           'exp': {'essential': True},
+           'iss': {'essential': True, 'value': issuer}
+           'aud': {"essential": True, "value": audience}
```
Comment (gao-sun) [inline docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx:151]:
```suggestion
           'aud': {'essential': True, 'value': audience}
```

File: docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx [hunk 3]
```diff
@@ -128,5 +128,56 @@ Package selections vary depending on the framework you are using.
...
+    def __init__(self, jwks, issuer="https://<your-logto-domain>/oidc"):
+        public_key = JsonWebKey.import_key_set(jwks)
+        super(MyLogtoBearerTokenValidator, self).__init__(public_key, issuer)
+        self.claims_options = {
+           'exp': {'essential': True},
+           'iss': {'essential': True, 'value': issuer}
+        }
+    def validate_token(self, token, scopes, request):
+        super(MyLogtoBearerTokenValidator, self).validate_token(token, scopes, request)
+        if token.get('aud') != str(request.url)
```
Comment (gao-sun) [inline docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx:177]: I think we can combine these two examples, and mark this function as optional? should be neater

File: docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx [hunk 4]
```diff
@@ -128,5 +128,56 @@ Package selections vary depending on the framework you are using.
...
+</TabItem>
+<TabItem value="python" label="Python">
+
+Leverage authlib's [JWTBearerTokenValidator](https://github.com/lepture/authlib/blob/169c7dcfc47478c8d55553cc95fb0f5578162b77/authlib/oauth2/rfc7523/validator.py)
+
+```python
+from authlib.oauth2.rfc7523 import JWTBearerTokenValidator
+from authlib.jose.rfc7517.jwk import JsonWebKey
+
+
```
Comment (gao-sun) [inline docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx:140]:
```suggestion

```

File: docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx (hunk 1 above)
Comment (simeng-li) [reply docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx:82]: Shouldn't matter. I'll update

File: docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx (hunk 2 above)
Comment (simeng-li) [reply docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx]: updated

File: docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx (hunk 3 above)
Comment (simeng-li) [reply docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx:158]: updated

File: docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx [hunk 5]
```diff
@@ -56,5 +56,39 @@ public class JwtRequestFilter extends OncePerRequestFilter {
 ```
 
 </TabItem>
+<TabItem value="python" label="Python">
 
+For python developers, there is a handy library [authlib](https://github.com/lepture/authlib) that can easily helps you to validate the breaer token issued by Logto.
```
Comment (charIeszhao) [inline docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx:61]:
```suggestion
For python developers, there is a handy library [authlib](https://github.com/lepture/authlib) that can easily help you validate the bearer token issued by Logto.
```

File: docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx [hunk 6]
```diff
@@ -56,5 +56,39 @@ public class JwtRequestFilter extends OncePerRequestFilter {
...
+```python
+from flask import Flask
+
+# For django developers:
+from authlib.integrations.django_oauth2 import ResourceProtector
+
+# For Flask Developers
+from authlib.integrations.flask_oauth2 import ResourceProtector
+
+# We will get back to this validator customization on the following validation section
```
Comment (charIeszhao) [inline docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx:79]:
```suggestion
# We will get back to this validator customization in the following validation section
```

File: docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx [hunk 7]
```diff
@@ -56,5 +56,39 @@ public class JwtRequestFilter extends OncePerRequestFilter {
...
+from validator import MyLogtoBearerTokenValidator
+
+app = Flask(__name__)
+
+require_auth = ResourceProtector()
+require_oauth.register_token_validator(MyLogtoBearerTokenValidator())
+
+@app.route("/api/foo")
+@require_auth()
+def foo() # your private API here
```
Comment (charIeszhao) [inline docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx:89]:
```suggestion
def foo() # your protected API here
```

**Review by charIeszhao — CHANGES_REQUESTED** [review]

**Review by charIeszhao — APPROVED** [review]
Overall looks good to me. But please check my nit comments

**Review by wangsijie — APPROVED** [review]

File: docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx [hunk 8]
```diff
@@ -128,5 +128,37 @@ Package selections vary depending on the framework you are using.
...
+           'iss': {'essential': True, 'value': issuer}
+           'aud': {'essential': True, "value": audience} # Optional if you are using request url to validate the token audience
+        }
+
+    """
+    OPTIONAL: To guard the token audience more strictly at each endpoint level:
+    """
+    def validate_token(self, token, scopes, request):
+        super(MyLogtoBearerTokenValidator, self).validate_token(token, scopes, request)
+        if token.get('aud') != str(request.url)
```
Comment (darcyYe) [inline docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx:158]: Missing tailing ":"?

File: docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx (hunk 5 above)
Comment (simeng-li) [reply docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx]: updated

File: docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx (hunk 7 above)
Comment (simeng-li) [reply docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx]: updated

File: docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx (hunk 6 above)
Comment (simeng-li) [reply docs/docs/recipes/protect-your-api/fragments/_token_extract.mdx]: updated

File: docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx (hunk 8 above)
Comment (simeng-li) [reply docs/docs/recipes/protect-your-api/fragments/_token_validation.mdx]: yep added

**Pull request merged**